Mathlib/Data/Fin/Basic.lean

```diff
@@ -244,6 +244,12 @@ This one instead uses a `NeZero n` typeclass hypothesis.
 theorem val_zero' (n : ℕ) [NeZero n] : ((0 : Fin n) : ℕ) = 0 :=
   rfl
 
+/-- Fin.mk_zero` in `Lean` only applies in `Fin (n + 1)`.
+This one instead uses a `NeZero n` typeclass hypothesis.
+-/
+@[simp]
+theorem mk_zero' (n : ℕ) (hn : 0 < n) : (⟨0, hn⟩ : Fin n) = haveI : NeZero n := .of_pos hn; 0 := rfl
```

Comment:
I'm not sure whether we want this `@[simp]` lemma to apply to any `(⟨0, hn⟩ : Fin n)` (as it is written now) or only if `[NeZero n]` is in the context. In the former case (as it's written now), it forces us to write LHS of other `simp` lemmas as `haveI : NeZero expr := _; ... (0 : Fin expr)` instead of `(⟨0, _⟩ : Fin expr)`.

Reply:
@kim-em WDYT? This lemma (in this form or in the `(n : Nat) [NeZero n]` form) is an obvious candidate for replacing `Fin.mk_zero` in the Lean core. Which form has more chances to get accepted?

Reply:
I think I'd prefer the `NeZero` version.

Reply:
Both versions have `NeZero`, just at different places. In one case, `simp` can lead to expressions that use `NeZero` instances that aren't in the context. In another case, adding a local `NeZero` instance may break `simp` calls.

Reply:
I went through `git grep 'haveI.*NeZero'`, and found that `Fin.cast_zero` and `PNat.mk_ofNat` are the only lemmas that introduce an inline `NeZero` instance to state the lemma.

Reply:
I've created #21229 which assumes `[NeZero _]` instead of adding inline instances. CC: @kim-em